node/src/main/scala/coop/rchain/node/diagnostics/effects/package.scala

```diff
@@ -61,26 +76,38 @@ package object effects {
       import A._
       import Trace._
 
+      private val spans: TrieMap[TraceId, SourceTrace] = TrieMap.empty
+
       override def mark(name: String): F[Unit] = {
-        ask.flatMap {
-          case DefaultTrace(source) =>
-            trace(source) {
-              mark(name)
-            }
-          case KamonTrace(ks) => KamonTracer.mark(ks, name)
+        ask.flatMap { traceId =>
+          val k: Option[Trace] = spans.get(traceId)
+          k match {
+            case Some(st: SourceTrace) => st.mark(name)
+            case _                     => Sync[F].unit
+          }
         }
       }.as(())
 
       override def trace[A](source: Source)(block: F[A]): F[A] =
         for {
-          span <- ask
-          r <- Sync[F].bracket((span match {
-                case DefaultTrace(_) => KamonTracer.start[F](source, networkId, host)
-                case KamonTrace(ks)  => KamonTracer.start[F](source, ks, networkId, host)
-              }).map(kamon)) {
-                scope(_)(block)
-              } {
-                case KamonTrace(ks) => KamonTracer.end(ks)
+          r <- Sync[F].bracket(ask.map { parentTraceId =>
+                val parent  = spans.get(parentTraceId)
+                val traceId = UUID.randomUUID()
+                spans
+                  .putIfAbsent(traceId, SourceTrace(source, parent))
+                  .map(_ => traceId)
+                  .getOrElse(traceId)
```

Comment:
shouldn't the span never be absent, if there's no bugs?
If so, let's raise an error if they're absent.